File: CONCURRENTE PARCIAL/client2.py
import socket
import threading
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import StandardScaler

class Person:
    def __init__(self, id, edad, sexo, peso, altura):
        self.id = id
        self.edad = edad
        self.sexo = sexo
        self.peso = peso
        self.altura = altura

class Client:
    def __init__(self, host, port):
        self.host = host
        self.port = port
        self.persons = []
        self.centroids = []
        self.cluster = []

    def run(self):
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as client:
                client.connect((self.host, self.port))
                print(f"Client has connected to server on port {self.port}")
                receiveDataThread = threading.Thread(target=self.receiveData, args=(client,))
                receiveDataThread.start()
        except Exception as e:
            print(e)

    def receiveData(self, client):
        try:
            with client:
                scanner = client.makefile(mode="r")
                bDataReceived = False
                while True:
                    if not bDataReceived:
                        message = scanner.readline().strip()
                        if message:
                            data = message.split("%")
                            personsString = data[0].split(";")
                            centroidsString = data[1].split(";")
                            self.parseData(personsString, self.persons)
                            self.parseData(centroidsString, self.centroids)
                            bDataReceived = True
                            self.calculateMortalityRisk(client)
                    else:
                        message = scanner.readline().strip()
                        if message:
                            centroidsString = message.split(";")
                            self.parseData(centroidsString, self.centroids)
                            self.calculateMortalityRisk(client)
        except Exception as e:
            print(e)

    def parseData(self, dataString, dataList):
        dataList.clear()
        for personString in dataString:
            personData = personString.strip('()\n').split(',')
            person = Person(int(personData[0]), int(personData[1]), int(personData[2]), int(personData[3]), int(personData[4]))
            dataList.append(person)

    def calculateMortalityRisk(self, client):
        # Prepare data for training the decision tree
        X = [[p.edad, p.sexo, p.peso, p.altura] for p in self.persons]
        y = [0] * len(self.persons)  # 0 for low risk, 1 for high risk

        # Train the decision tree
        sc = StandardScaler()
        X = sc.fit_transform(X)
        clf = DecisionTreeClassifier(random_state=42)
        clf.fit(X, y)

        # Predict the risk for each person
        for person in self.persons:
            risk = clf.predict([[person.edad, person.sexo, person.peso, person.altura]])
            self.cluster.append(risk[0])

        # Send the results to the server
        message = "[" + ", ".join(str(x) for x in self.cluster) + "]" + "\n"
        client.send(message.encode())
        self.cluster.clear()
        print("Data sent to server")

if __name__ == "__main__":
    host = "localhost"
    port = 2206
    client = Client(host, port)
    client.run()

File: CONCURRENTE PARCIAL/Client2.java
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.Scanner;
import java.util.Vector;

public class Client2 {
    private static final int port = 2206;
    private static final String host = "localhost";
    private static Socket client;

    private static final Vector<Persona> personas = new Vector<>();
    private static final MessageQueue resultQueue = new MessageQueue();

    public static void main(String[] args){
        try {
            client = new Socket(host, port);
            System.out.println("Client has connected to server on port " + client.getPort());
            Thread receiveDataThread = new Thread(() -> {
                try (Scanner scanner = new Scanner(client.getInputStream())) {
                    boolean bDataReceived = false;
                    while (true) {
                        if (!bDataReceived) {
                            if (scanner.hasNextLine()) {                                
                                    String message = scanner.nextLine();
                                    System.out.println(message);
                                    String[] data = message.split(";");
                                    bDataReceived = true;
                                    parseData(data);           
                            }
                        }else{
                            if(scanner.hasNextLine()){
                                String message = scanner.nextLine();
                                String[] data = message.split(";");
                                parseData(data);                               
                            }
                        }
                    }
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            });
            receiveDataThread.start();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }   
    
    private static void parseData(String[] data) {
        try{

            for (String datum : data) {
                String[] attributes = datum.split(",");
                int id = Integer.parseInt(attributes[0].substring(1));
                int edad = Integer.parseInt(attributes[1]);
                int sexo = Integer.parseInt(attributes[2]);
                int peso = Integer.parseInt(attributes[3]);
                int altura = Integer.parseInt(attributes[4].substring(0, attributes[4].length() - 1));
                //System.out.println(id+""+edad+""+sexo+""+peso+""+altura);
                personas.add(new Persona(id, edad, sexo, peso, altura));
            }

        } catch (Exception e){
            System.out.println(e);
        }
        System.out.println("PROCESS: ");
        processData();
    }

    private static void processData() {
        try {
            System.out.println("Processing data...");
            DecisionTree tree = new DecisionTree();
            for (Persona persona : personas) {
                int sexo = persona.getSexo();
                int edad = persona.getEdad();
                int peso = persona.getPeso();
                int altura = persona.getAltura();
                int mortality = tree.predict(sexo, edad, peso, altura);
                System.out.println("mortalidad: "+mortality);
                resultQueue.addMessage(String.valueOf(mortality));
            }
            System.out.println("Data processed successfully.");
        } catch (Exception e) {
            e.printStackTrace();
        }
        sendResults();
    }

    private static void sendResults() {
        try {
            System.out.println("Sending results...");
            
            OutputStream outputStream = client.getOutputStream();
            while (!resultQueue.isEmpty()) {
                //System.out.println(resultQueue.getNextMessage());
                String result = resultQueue.getNextMessage() + "\n";
                outputStream.write(result.getBytes());
                outputStream.flush();
            }
            System.out.println("Results sent successfully.");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
